torch/csrc/distributed/c10d/comm.cpp

```diff
@@ -87,6 +87,14 @@ const std::vector<at::Tensor>& GradBucket::getTensors() {
   return tensors_;
 }
 
+const at::Tensor& GradBucket::getTensor() {
+  TORCH_CHECK(
+      tensors_.size() == 1,
+      "GradBucket::getTensor can only be used when tensors "
+      "contain bucket contents tensor for just one replica.")
```

Comment:
Does this mean that the comm. hook does not work when modules are replicated across GPU in a single process (single-process, multi-device)? This is fine as we're retiring support for that, but would be good to confirm and maybe clarify in docs (cc @mrshenli )

Reply:
I removed `getTensor` after the thread in https://github.com/pytorch/pytorch/pull/42335#discussion_r463769165  
I think `process_group.allreduce()` gets list of Tensors and it should support the multiple replica case. 

Looks like we need to `allreduce` separately for each replica in the case of `dist.all_reduce()`. Since we need to combine all these `allreduce` into one Future, I'm not really sure how the hook will look like for that case though.

Reply:
> I'm not really sure how the hook will look like for that case though.

in that case, we can either:

1. add a warning to state that the hook does not work with SPMD mode
2. in `_register_comm_hook` detect if it is SPMD, if yes, error out.

Reply:
@mrshenli but `process_group.allreduce()` gets list of Tensors, so it should work with SPMD mode, right?

Reply:
@sinannasir yep, I agree. But as a framework, we don't promise on features without test coverage. If you plan to imply it would work for both SPMD DDP and SPSD DDP, we will need to add tests for both modes.

Reply:
@sinannasir To keep this PR simple, let's just error out in the SPMD mode and create a gh issue to address this later.